### PR TITLE
chore: StateMachine tests only bump time if strictly necessary

### PR DESCRIPTION
**Pull request by mraszyk**
This PR makes StateMachine tests only bump time by 1ns before executing a round if strictly necessary, i.e., if time has not increased since the last round.

### CODE REVIEW

**Review by mbjorkqvist — APPROVED** [review]

**Comment by berestovskyy** [comment]
Hey folks, the exact time of the next round is currently unknown in production. Can someone clarify the specific use case for the tests requiring this information?

**Comment by mraszyk** [comment]
> Hey folks, the exact time of the next round is currently unknown in production. Can someone clarify the specific use case for the tests requiring this information?

Tests use that information to know what time to expect an update call will be evaluated on. In some cases, this could be achieved by making that update call and checking the current time afterwards. However, if the update call makes downstream calls, then the current time after evaluating the update call might be larger than the time at which the update call started to process.

Please speak up if there are other use cases of that information in your tests!

**Review by aterga — APPROVED** [review]

**Comment by berestovskyy** [comment]
> However, if the update call makes downstream calls, then the current time after evaluating the update call might be larger than the time at which the update call started to process.

My point is that we should maybe not assert the exact time, as in production we wouldn't guess it.

Anyway, up to you...

**Review by berestovskyy — APPROVED** [review]

File: rs/state_machine_tests/src/lib.rs [hunk 1]
```diff
@@ -2442,9 +2446,16 @@ impl StateMachine {
         SystemTime::UNIX_EPOCH + Duration::from_nanos(self.time.load(Ordering::Relaxed))
     }
 
-    /// Returns the state machine time at the beginning of next round.
+    /// Returns the state machine time at the beginning of next round
+    /// under the assumption that time won't be advanced before that
+    /// round is finalized.
     pub fn time_of_next_round(&self) -> SystemTime {
```
Comment (schneiderstefan) [inline rs/state_machine_tests/src/lib.rs:2452]: Side note: The fact that there is a `time_of_next_round` and a different `get_time_of_next_round` just below is a bit weird.

File: rs/state_machine_tests/src/lib.rs [hunk 2]
```diff
@@ -847,6 +847,7 @@ pub struct StateMachine {
     /// A drop guard to gracefully cancel the ingress watcher task.
     _ingress_watcher_drop_guard: tokio_util::sync::DropGuard,
     query_stats_payload_builder: Arc<PocketQueryStatsPayloadBuilderImpl>,
+    time_of_last_round: RwLock<SystemTime>,
```
Comment (schneiderstefan) [inline rs/state_machine_tests/src/lib.rs:850]: Could you explain to me the relationship of the fields `time` and `time_of_last_round` and the two getters that use them, namely `time()` and `time_of_next_round()`?

File: rs/state_machine_tests/src/lib.rs (hunk 1 above)
Comment (mraszyk) [reply rs/state_machine_tests/src/lib.rs:2452]: Is it confusing that there are two function calls or that there are calls to functions with different names (or both)?

File: rs/state_machine_tests/src/lib.rs (hunk 2 above)
Comment (mraszyk) [reply rs/state_machine_tests/src/lib.rs:850]: The `time` is the time that will determine the time of the next round and users can advance it. The `time_of_last_round` is the time of the last executed round.

Comment (schneiderstefan) [reply rs/state_machine_tests/src/lib.rs:850]: Ah ok I think I get it now. Could you maybe move `time_of_last_round` next to `time` and add a comment?

File: rs/state_machine_tests/src/lib.rs (hunk 1 above)
Comment (schneiderstefan) [reply rs/state_machine_tests/src/lib.rs:2452]: Mainly that there are two functions with almost the same name and almost the same functionality, except the return type. But this isn't something that needs changing, I was just confused at first.

File: rs/state_machine_tests/src/lib.rs (hunk 2 above)
Comment (mraszyk) [reply rs/state_machine_tests/src/lib.rs]: Done.

**Review by fxgst — APPROVED** [review]

**Pull request merged**